clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -1291,6 +1294,10 @@ private void releaseAssignmentAndLeaveGroup(final Timer timer) {
             log.warn("Consumer triggered an unsubscribe event to leave the group but couldn't " +
                 "complete it within {} ms. It will proceed to close.", timer.timeoutMs());
         } finally {
+            // Regardless of success or failure of the unsubscribe process, it's important to process any background
+            // events in the hope that our ConsumerRebalanceListenerCallbackNeededEvent is present and can be executed.
+            processBackgroundEvents();
```

Comment:
With this PR we're addressing not only the interrupt, but also the close(0). If I get it right, this line is what makes that the close(0) will always send the leave request (even if there are callbacks), correct? Could we add then an IT to see how close(0) with callbacks makes that the consumer leaves the group? (I expect it should be very similar to the IT added for the interrupt case, but no interrupt + close(0) )

Reply:
Yes. Good idea!

Reply:
are we adding this test?

Reply:
just for the record, my expectations here were not right because of another gap discovered. This `processBackgrounEvents` is not enough to ensure that close(0) sends the leave, and https://issues.apache.org/jira/browse/KAFKA-17518 was created.

Reply:
Among other things, the background thread uses background events to signal the application thread to execute the `ConsumerRebalanceListener` callbacks.

It was my understanding from looking at the code for the existing `ClassicKafkaConsumer` and from previous discussions on this topic that we need to execute the `ConsumerRebalanceListener.onPartitionsRevoked()` method when unsubscribing/leaving a group.